Dev/heat usm

rscohn2 opened the pull request:
# Description

Adds USM to 1dheat sample. First commit comes from another PR: #251

anjgola commented:
As we address the errors in a separate thread..in the meantime
1. Requesting @tomlenth to review for DX/readability standards.
2. Requesting @lqnguyen to review as the original author in the sample.

tomlenth commented:
A few minor comments: 

if "$" are included in the commands in step one, they should also be included in step 2 and 3.

 I'm confused by the application parameters.  
- Are the parameters passed via command line? if so, how.  or, does user modify in source?  Just a simple explanation or example how to use would be great. 
-the formatting of the parameters description is odd, making it look like a typo, esp given the missing capitalization and punctuation.  try the following: 

Where:
- n is the number of points you want to simulate the heat transfer.
- i is the number of timesteps in the simulation.

lqnguyen commented:
A quick look into the proposed change, following are my comments:
1. Keep the allocation of device memory in main instead of moving to ComputeHeatDevice. Keep it in the main is easier to compare with the CPU memory.
2. When calling the function CompareResults, we should check the return Boolean so that we know if there is an error in parallel_error_diff.txt or not. 
3. Somehow, the proposed implementation is not correct, please take a look in  the output file: parallel_error_diff.txt. All the numbers are "0".

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+// Initialize the temperature arrays
+//
+void Initialize(float* arr, float* arr_next, unsigned int num) {
+  arr[0] = arr_next[0] = initial_temperature;
```
This looks wrong, the arr[0]=.... line should be last to not be overwritten by the loop.  Or loop start from 1.

rscohn2 replied:
Good catch. Initializing the array to all 0 masked my other bugs.

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
 
-  try {
-    // Define the device queue
-    queue q = default_selector{};
```
Should follow dpc_common.hpp pattern from async exceptions.

rscohn2 replied:
fixed.
Note that dpc_common.hpp is not part of open source dpc++ so the samples are not portable.

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+
+	       auto step =
+		 [=](id<1> k) {
+		   
```
Extra blank line.

rscohn2 replied:
Both lines?

pmpeter1 replied:
I was talking about line 138.   The blank line at 135 is more normal to separate declarations from code.

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+//
+void Initialize(float* arr, float* arr_next, unsigned int num) {
+  arr[0] = arr_next[0] = initial_temperature;
+  for (unsigned int i = 0; i < num; i++)
```
Should be i=1 to not overwrite initial_temperature.

rscohn2 replied:
Fixed

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+void ComputeHeatDevice(float C, unsigned int num_p, unsigned int num_iter,
+		       float* arr_CPU) {
+  // Define the device queue
+  queue q;
```
Use dpc_common async exception handler

rscohn2 replied:
Fixed

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+			  float* arr_CPU) {
+  // Use an in-order queue so the time steps are executed in-order
+  sycl::property_list properties{sycl::property::queue::in_order()};
+  queue q(properties);
```
Use dpc_common async exception handler.

rscohn2 replied:
fixed

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+    q.submit([&](auto& h) {
+	       auto step =
+		 [=](id<1> k) {
+		   
```
Extra blank line

rscohn2 replied:
fixed

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+
+  CompareResults("parallel", arr_next, arr_CPU, num_p, C);
+
+  sycl::free(arr, q);
```
Why use sycl:: prefix here?

rscohn2 replied:
removed

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+void ComputeHeatDeviceUSM(float C, unsigned int num_p, unsigned int num_iter,
+			  float* arr_CPU) {
+  // Use an in-order queue so the time steps are executed in-order
+  sycl::property_list properties{sycl::property::queue::in_order()};
```
Why use sycl:: here?

rscohn2 replied:
removed

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+	     });    // end device queue
+    
+    // Swap arrarys for next step
+    float* swap = arr;
```
Use consistent code for swap.   Maybe switch to std::swap?   https://en.cppreference.com/w/cpp/algorithm/swap

rscohn2 replied:
Using std::swap

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -177,50 +240,18 @@ float* ComputeHeatHostSerial(float* arr, float* arr_next, float C,
-  unsigned int n_point;  // The number of point in 1D space
-  unsigned int
-      n_iteration;  // The number of iteration to simulate the heat propagation
+  unsigned int n_point;      // The number of point in 1D space
```
Would be more concise to use size_t instead of unsigned int.   https://en.cppreference.com/w/cpp/types/size_t

rscohn2 replied:
Changed to size_t

rscohn2 commented:
> I'm confused by the application parameters.
> 
> * Are the parameters passed via command line? if so, how.  or, does user modify in source?  Just a simple explanation or example how to use would be great.
>   -the formatting of the parameters description is odd, making it look like a typo, esp given the missing capitalization and punctuation.  try the following:
> 
> Where:
> 
> * n is the number of points you want to simulate the heat transfer.
> * i is the number of timesteps in the simulation.

Added example usage and your suggested change

anjgola reviewed (approved):
Start CI.

rscohn2 commented:
> A quick look into the proposed change, following are my comments:
> 
> 1. Keep the allocation of device memory in main instead of moving to ComputeHeatDevice. Keep it in the main is easier to compare with the CPU memory.

I like to keep main as small as possible. I moved the compare down into the functions that execute the kernels.

> 2. When calling the function CompareResults, we should check the return Boolean so that we know if there is an error in parallel_error_diff.txt or not.
I moved the check into the compare function, and eliminated the return value.
> 3. Somehow, the proposed implementation is not correct, please take a look in  the output file: parallel_error_diff.txt. All the numbers are "0".

 I fixed several errors, including the one that was masking the fails.

rscohn2 commented:
@pmpeter1 @anjgola @tomlenth @lqnguyen 
I have responded to all review comments and the code now passes CI, so please take a look. I made many changes since review so I recommend scanning through the code versus looking at the extra commits. Please 'squash and merge' if we are done.

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -47,11 +48,13 @@ using namespace std;
+//
+// Display input parameters used for this sample.
+//
 void Usage(string programName) {
```
Might be better at const string &programName

rscohn2 replied:
I changed it

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,173 +63,208 @@ void Usage(string programName) {
+
+  for (size_t i = 0; i < num_point + 2; i++) {
+    err_file << "\n RESULT: " << i << "\t" << std::setw(12) << std::left
+             << device_results[i] << "\t" << host_results[i];
```
Is there a missing "\n" here?

rscohn2 replied:
I moved the code, but did not write it. There is a \n at the beginning. I moved it to the end.

pmpeter1 commented on DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp:
```diff
@@ -60,173 +63,208 @@ void Usage(string programName) {
 
   // Read input parameters
   try {
     n_point = stoi(argv[1]);
```
So you need to worry about the user passing in -1 as an argument?

rscohn2 replied:
I added some error checking.

rscohn2 replied:
Old code did not handle -1. I added some error checking.

tomlenth reviewed (approved):
thanks, looks good

rscohn2 commented:
@pmpeter1 Can you merge? I am working on multi-gpu variants in parallel and you will have opportunities for more changes with that commit. The more we delay, the more they diverge.

anjgola commented:
Waiting for 'no merge period' to end.
